core/src/main/java/org/elasticsearch/action/index/TransportIndexAction.java

```diff
@@ -175,29 +182,40 @@ protected Location onReplicaShard(IndexRequest request, IndexShard indexShard) {
     }
 
     public static WriteResult<IndexResponse> executeIndexRequestOnPrimary(IndexRequest request, IndexShard indexShard,
-            MappingUpdatedAction mappingUpdatedAction) throws Exception {
-        Engine.Index operation = prepareIndexOperationOnPrimary(request, indexShard);
+            MappingUpdatedAction mappingUpdatedAction) {
+        Engine.Index operation;
+        try {
+            operation = prepareIndexOperationOnPrimary(request, indexShard);
+        } catch (Exception e) {
+            return new WriteResult<>(e);
+        }
         Mapping update = operation.parsedDoc().dynamicMappingsUpdate();
         final ShardId shardId = indexShard.shardId();
         if (update != null) {
-            mappingUpdatedAction.updateMappingOnMaster(shardId.getIndex(), request.type(), update);
-            operation = prepareIndexOperationOnPrimary(request, indexShard);
+            try {
+                mappingUpdatedAction.updateMappingOnMaster(shardId.getIndex(), request.type(), update);
+                operation = prepareIndexOperationOnPrimary(request, indexShard);
+            } catch (Exception e) {
```

Comment:
Now we let these failures bubble up to be retried by the replication operation. Initially, I thought it would be a good idea to handle these failures explicitly by inspecting the write result at the replication operation layer, now I see it makes more sense to let replication operation deal with these as is and only capture the operation-level failures to be dwelt explicitly